File: Top_100_Liked_Questions/78._Subsets/solution.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        # cascading method
        ans = [[]]
        for n in nums:
            for subset in ans[:]:
                ans.append(subset + [n])
        
        return ans
    
    def subsets2(self, nums: List[int]) -> List[List[int]]:
        # using binary
        ans = []
        for dec in range(2**len(nums)):
            b = bin(dec)[2:]
            s = ('0' * (len(nums) - len(b))) + b
            subset = []
            for i,v in enumerate(s):
                if v == '1':
                    subset.append(nums[i])
            ans.append(subset)
        
        return ans
                    

File: Top_100_Liked_Questions/78._Subsets/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    // cascading method
    let ans = [[]];
    for (const n of nums) {
        for (let subset of [...ans]) {
            ans.push([...subset, n]);
        }
    }
    return ans;
}

var subsets2 = function(nums) {
    // using binary
    let ans = [];
    
    for (let i=0; i<Math.pow(2,nums.length); i++) {
        let bin = i.toString(2);
        let zeroFilledBin = '0'.repeat(nums.length - bin.length) + bin;
        let subset = []
        for (let i=0; i<zeroFilledBin.length; i++) {
            if (zeroFilledBin[i] === '1') {
                subset.push(nums[i]);
            }
        }
        ans.push(subset);
    }
    
    return ans;
};